### PR TITLE
Return digest value

### PR DESCRIPTION
**Pull request by unmultimedio**

### CODE REVIEW

File: private/pkg/manifest/manifest_test.go [hunk 1]
```diff
@@ -83,14 +83,14 @@ func TestAddEntry(t *testing.T) {
...
+	require.NoError(t, m.AddEntry(filePath, fileDigest)) // adding the same entry twice is fine
 
-	require.Error(t, m.AddEntry("", *fileDigest))
+	require.Error(t, m.AddEntry("", fileDigest))
 	require.Error(t, m.AddEntry("other/path", manifest.Digest{}))
-	require.Error(t, m.AddEntry(filePath, *mustDigestShake256(t, []byte("other content"))))
+	require.Error(t, m.AddEntry(filePath, mustDigestShake256(t, []byte("other content"))))
 
-	expect := fmt.Sprintf("%s  %s\n", fileDigest, filePath)
+	expect := fmt.Sprintf("%s  %s\n", fileDigest.String(), filePath)
```
Comment (unmultimedio) [inline private/pkg/manifest/manifest_test.go:93]: This is downside 1, is not an straight forward stringer.

File: private/pkg/manifest/digest_test.go [hunk 2]
```diff
@@ -29,17 +29,18 @@ import (
 
 func TestNewDigestBytes(t *testing.T) {
 	t.Parallel()
+	nilDigest := mustDigestShake256(t, nil)
```
Comment (unmultimedio) [inline private/pkg/manifest/digest_test.go:32]: This is downside 2, cannot chain to get `.Bytes()`, or `.Type()`, or similar.

File: private/pkg/manifest/module.go [hunk 3]
```diff
@@ -35,7 +35,7 @@ var (
 
 // Blob is a blob with a digest and a content.
 type Blob interface {
-	Digest() *Digest
+	Digest() Digest
```
Comment (unmultimedio) [inline private/pkg/manifest/module.go:38]: This PR is just this change.

File: private/pkg/manifest/module.go [hunk 4]
```diff
@@ -169,7 +170,8 @@ func NewBlobSet(ctx context.Context, blobs []Blob, opts ...BlobSetOption) (*Blob
 	}
 	digestToBlobs := make(map[string]Blob, len(blobs))
 	for _, b := range blobs {
-		digestStr := b.Digest().String()
+		digest := b.Digest()
+		digestStr := digest.String()
```
Comment (unmultimedio) [inline private/pkg/manifest/module.go:174]: From downside 1: this sometimes could be a bit verbose

**Review by unmultimedio — COMMENTED** [review]
Only 2 downsides I see, the rest are wins imo.

File: private/pkg/manifest/manifest_test.go (hunk 1 above)
Comment (unmultimedio) [reply private/pkg/manifest/manifest_test.go:93]: Although it could be fixed passing the pointer 🤔

File: private/pkg/manifest/module.go (hunk 3 above)
Comment (twilly) [reply private/pkg/manifest/module.go:38]: And there's two implicit downsides here. One is Digests on the heap would be copied to stack. The other is a blob that doesn't have a digest must produce one. That latter is why chaining becomes possible: it requires blobs to have digests. Downside or upside is a matter of perspective with that one.

File: private/pkg/manifest/module.go (hunk 4 above)
Comment (twilly) [reply private/pkg/manifest/module.go:174]: Digests has pointer receivers. If we really wanted to return Digest as values, I think we should be considering Digest values as immutable and define only value receivers. Mixing the two is causing headaches like these.

**Comment by unmultimedio** [comment]
Converting to draft, wait to see if it's really needed.

**Comment by unmultimedio** [comment]
Closing now, will reopen it if shows that pointers receivers will be a greater issue.

**Pull request closed without merging**